Moves and updates basic development workflow

lukpueh opened the pull request:
@awwad has created a nice document about our lab development workflow in https://github.com/secure-systems-lab/code-style-guidelines/pull/8. I think this repo is better fit for such instructions. So after plenty of rebasing and cherry picking I moved his document over here.

While doing that I updated the basic workflow and stripped out some WIP *further reading* sections and added them to a new document available in the [dev-workflow-misc branch](https://github.com/secure-systems-lab/lab-guidelines/tree/dev-workflow-misc), which still needs some work.

I think for now, we should be fine with this basic description.  But, please feel free to review, comment and/or change, ...

awwad committed:
Add skeleton for lab development guidelines

awwad committed:
Outline general workflow and link to commit best practices

awwad committed:
Link to Seattle Git and GitHub guide

awwad committed:
Expand commit message guidelines with rules and examples

awwad committed:
Add high-level PR rules

awwad committed:
Reorganize sections / heading levels / titles

vladimir-v-diaz committed:
Fix typo in section "General Development Workflow in the Lab"

awwad committed:
Link to directions on creating pull request templates

Commit:
Rename development_workflow.md for fn conistency

This repo seems to have hyphenated file names

Commit:
Polish development guidelines intro and workflow

Add a little intro emphasizing the importance of git and GitHub
for our lab and polish the general development workflow as prepared
by @awwad.

Commit:
Remove extra topics from development workflow

These topics will be revised and merged with
lab-guidelines@dev-workflow-misc.

awwad committed:
Grammar, typos, and spacing

awwad committed:
Clarify and add separate step for code style guidelines

awwad committed:
Add minimal commit and pull request guidelines doc

awwad committed:
Minor grammatical tweaks

awwad commented:
I hijacked this a bunch, as discussed. 🙊 

What do you think of the direction taken, @lukpueh ?

Commit:
Minor rewording and formatting changes

- Replace tex-style quotes with verbatim backticks to clarify
  that contributors should label their unfinished commits
  "(WIP)" (without quotes)
- Change "commit" section title to not confuse it with the
  more advanced commit splitting, e.g. when rebasing interactively
- Emphasize "git" and "GitHub" mentions using italic markup
- Markup a git command with verbatim backticks

Commit:
Add WIP label to "PR guidelines" section title

This section is missing some contents. The existing contents are
nonetheless accurate.

lukpueh commented:
I would say the overall workflow in  `dev-workflow.md` and the commit section in `commits.md` are pretty solid. Let's merge and create a separate PR for an update of the *"Pull Request guidelines"* which I have labeled *(WIP)* for now. What does everyone think?

vladimir-v-diaz commented:
I think it's fine to merge.

Is there a reason why this pull request must be merged now?

lukpueh commented:
The reason why I'd like to merge now is that I've repeatedly pointed students and co-workers to this PR (because IMHO it already gives an excellent overview of our typical workflow). 
So I thought merging would acknowledge that these guidelines are already valid. And it would also give us a base for all sorts of enhancements, e.g. [more details on PRs](a8ed96e62493e5f11bdd49d96d67abaab6ddf70a), [linking to GitHub templates](https://github.com/secure-systems-lab/github-templates), [including git and GitHub tutorials](https://github.com/secure-systems-lab/code-style-guidelines/pull/10), ...

aaaaalbert commented on commits.md:
```diff
@@ -0,0 +1,125 @@
+file format). You may even want to keep changes that different on distinct
+branches and make distinct pull requests for them.
+
+#### Separate Moving and Modifying Files
```
I don't think GitHub has the describe shortcomings actually. I happened to rename and edit a file at some point, and the GitHub diff looks just fine, and the diff box header shows the rename: https://github.com/SeattleTestbed/repy_v2/pull/134/files#diff-08d644fdf23e49274b1b1fb57fb15bb1 .

(Also, I'm not sure about unconditionally separating moves and edits for the other reasons given. The interdependency of not-yet reviewed PRs can be a drag.)

Commit:
Minor rephrasing in development guidelines

Commit:
Add reference to dev guideline in README

lukpueh commented:
I'd say we have something we can use. Let's merge and fix/add contents as we move forward.

vladimir-v-diaz commented:
I don't think the README links to the dev workflow.

lukpueh commented:
I think [it does](https://github.com/secure-systems-lab/lab-guidelines/blame/master/README.md#L16-L17).
